### PR TITLE
New spread-shellcheck tool for spread tasks

### PR DESCRIPTION
**Pull request by sergiocazzolato**
This change introduces the new spread-shellcheck tool and a test which
is validating all the tests.

### CODE REVIEW

File: tools/spread-shellcheck [hunk 1]
```diff
@@ -0,0 +1,397 @@
+#!/usr/bin/env python3
+
+# Copyright (C) 2018 Canonical Ltd
```
Comment (Meulengracht) [inline tools/spread-shellcheck]: This should be 2022 instead of 2018

File: tools/spread-shellcheck [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+    script_data = []
+    script_data.append('set -e')
+
+    for key, value in env.items():
+        value = str(value)
+        # Unpack the special "$(HOST: ...) syntax and tell shellcheck not to
+        # worry about the use of echo to print variable value.
+        if value.startswith("$(HOST:") and value.endswith(")"):
+            script_data.append("# shellcheck disable=SC2116")
+            value = "$({})".format(value[len("$(HOST:"):-1])
```
Comment (Meulengracht) [inline tools/spread-shellcheck:126]: Nitpick again: You can also use string interpolation here if you want (and in rest of file)

File: tools/spread-shellcheck [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+
+def findfiles(locations, exclude):
+    for loc in locations:
+        if os.path.isdir(loc):
+            for root, _, files in os.walk(loc, topdown=True):
+                for name in files:
+                    if name in ['spread.yaml', 'task.yaml']:
+                        full_path = os.path.join(root, name)
+                        if not is_file_in_dirs(full_path, exclude):
+                            yield full_path
```
Comment (Meulengracht) [inline tools/spread-shellcheck:287]: Why are we using yield here? I can see it's called from line 302, but I don't see it's called multiple times? This is probably correct usage, I'm just trying to understand how 'findfiles' is called multiple times to utilize the use of 'yield'

Or maybe I am completely wrong about how yield works in python

**Review by Meulengracht — COMMENTED** [review]
This is really nice! Thanks for doing this, it's really appreciated - I just have a few minor comments

File: tools/spread-shellcheck (hunk 3 above)
Comment (sergiocazzolato) [reply tools/spread-shellcheck:287]: Hi, thanks for the review, yield is used to create a generator in python, perhaps you are confused with shield. Using yield you define which are the elements for the generator.

File: tools/spread-shellcheck (hunk 2 above)
Comment (sergiocazzolato) [reply tools/spread-shellcheck:126]: @Meulengracht, The only problem I see is that interpolation was introduced in python3.6 and in trusty we have python 3.4 and in xenial we have python 3.5. 
Perhaps we should have this in a separate dir called utils, so utils are used not in the supported systems, but in a subset of systems where we run the workflows, so we can make sure in those instances we will have python 3.6.
Does it make sense?

File: tools/spread-shellcheck (hunk 1 above)
Comment (sergiocazzolato) [reply tools/spread-shellcheck]: done

File: tools/spread-shellcheck (hunk 2 above)
Comment (Meulengracht) [reply tools/spread-shellcheck:126]: @sergiocazzolato Ah that makes so much sense! Thanks for clarifying this for me, I was not aware of this.

File: tools/spread-shellcheck (hunk 3 above)
Comment (Meulengracht) [reply tools/spread-shellcheck:287]: Ah so the function returns a generator - not an element at each call. Thanks for explaining this.

**Review by Meulengracht — APPROVED** [review]
Thanks again, LGTM

**Pull request merged**